File: example/ios/WebViewBridgeDemo/WebViewBridgeDemo/CustomModule.swift
/**
 * Custom 模块 - 自定义模块示例
 *
 * 演示如何创建自定义原生模块，供开发者参考
 * 包含 Alert、Confirm、Prompt、Toast 等常见 UI 交互方法
 */

import UIKit
import WebViewBridge

/// 自定义模块 - 演示原生 UI 交互
///
/// 此模块展示如何创建自定义模块，包含：
/// - Alert: 显示原生警告框
/// - Confirm: 显示确认对话框
/// - Prompt: 显示输入对话框
/// - Toast: 显示 Toast 消息
/// - Loading: 显示/隐藏加载指示器
/// - ActionSheet: 显示操作表
class CustomModule: BridgeModule {
    
    let moduleName = "Custom"
    let methods = ["Alert", "Confirm", "Prompt", "Toast", "ShowLoading", "HideLoading", "ActionSheet"]
    
    private weak var viewController: UIViewController?
    
    /// Loading 遮罩视图
    private var loadingView: UIView?
    
    init(viewController: UIViewController) {
        self.viewController = viewController
    }
    
    func handleRequest(
        method: String,
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        switch method {
        case "Alert":
            showAlert(params: params, callback: callback)
        case "Confirm":
            showConfirm(params: params, callback: callback)
        case "Prompt":
            showPrompt(params: params, callback: callback)
        case "Toast":
            showToast(params: params, callback: callback)
        case "ShowLoading":
            showLoading(params: params, callback: callback)
        case "HideLoading":
            hideLoading(callback: callback)
        case "ActionSheet":
            showActionSheet(params: params, callback: callback)
        default:
            callback(.failure(BridgeError.methodNotFound("\(moduleName).\(method)")))
        }
    }
    
    // MARK: - Alert
    
    private func showAlert(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let title = params["title"]?.value as? String
        let message = params["message"]?.value as? String ?? ""
        let buttonText = params["buttonText"]?.value as? String ?? "确定"
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: buttonText, style: .default) { _ in
                callback(.success(["action": "confirm"]))
            })
            vc.present(alert, animated: true)
        }
    }
    
    // MARK: - Confirm
    
    private func showConfirm(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let title = params["title"]?.value as? String
        let message = params["message"]?.value as? String ?? ""
        let confirmText = params["confirmText"]?.value as? String ?? "确定"
        let cancelText = params["cancelText"]?.value as? String ?? "取消"
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: cancelText, style: .cancel) { _ in
                callback(.success(["confirmed": false]))
            })
            alert.addAction(UIAlertAction(title: confirmText, style: .default) { _ in
                callback(.success(["confirmed": true]))
            })
            vc.present(alert, animated: true)
        }
    }
    
    // MARK: - Prompt
    
    private func showPrompt(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let title = params["title"]?.value as? String
        let message = params["message"]?.value as? String
        let placeholder = params["placeholder"]?.value as? String ?? ""
        let defaultValue = params["defaultValue"]?.value as? String ?? ""
        let confirmText = params["confirmText"]?.value as? String ?? "确定"
        let cancelText = params["cancelText"]?.value as? String ?? "取消"
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            alert.addTextField { textField in
                textField.placeholder = placeholder
                textField.text = defaultValue
            }
            
            alert.addAction(UIAlertAction(title: cancelText, style: .cancel) { _ in
                callback(.success(["confirmed": false, "value": ""]))
            })
            alert.addAction(UIAlertAction(title: confirmText, style: .default) { _ in
                let value = alert.textFields?.first?.text ?? ""
                callback(.success(["confirmed": true, "value": value]))
            })
            vc.present(alert, animated: true)
        }
    }
    
    // MARK: - Toast
    
    private func showToast(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let message = params["message"]?.value as? String ?? ""
        let duration = params["duration"]?.value as? String ?? "short"
        let displayDuration: TimeInterval = duration == "long" ? 3.0 : 2.0
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            // 创建 Toast 视图
            let toastLabel = UILabel()
            toastLabel.backgroundColor = UIColor.black.withAlphaComponent(0.7)
            toastLabel.textColor = .white
            toastLabel.textAlignment = .center
            toastLabel.font = UIFont.systemFont(ofSize: 14)
            toastLabel.text = message
            toastLabel.alpha = 0
            toastLabel.layer.cornerRadius = 10
            toastLabel.clipsToBounds = true
            toastLabel.numberOfLines = 0
            
            let maxWidth = vc.view.bounds.width - 60
            let size = toastLabel.sizeThatFits(CGSize(width: maxWidth, height: .greatestFiniteMagnitude))
            let width = min(size.width + 32, maxWidth)
            let height = size.height + 16
            
            toastLabel.frame = CGRect(
                x: (vc.view.bounds.width - width) / 2,
                y: vc.view.bounds.height - 120,
                width: width,
                height: height
            )
            
            vc.view.addSubview(toastLabel)
            
            UIView.animate(withDuration: 0.3, animations: {
                toastLabel.alpha = 1.0
            }) { _ in
                UIView.animate(withDuration: 0.3, delay: displayDuration, options: [], animations: {
                    toastLabel.alpha = 0
                }) { _ in
                    toastLabel.removeFromSuperview()
                }
            }
            
            callback(.success(nil))
        }
    }
    
    // MARK: - Loading
    
    private func showLoading(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let message = params["message"]?.value as? String ?? "加载中..."
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            // 移除现有的 loading 视图
            self?.loadingView?.removeFromSuperview()
            
            // 创建遮罩
            let overlay = UIView(frame: vc.view.bounds)
            overlay.backgroundColor = UIColor.black.withAlphaComponent(0.4)
            overlay.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            
            // 创建容器
            let container = UIView()
            container.backgroundColor = UIColor.white
            container.layer.cornerRadius = 12
            container.translatesAutoresizingMaskIntoConstraints = false
            
            // 创建指示器
            let indicator = UIActivityIndicatorView(style: .large)
            indicator.translatesAutoresizingMaskIntoConstraints = false
            indicator.startAnimating()
            
            // 创建标签
            let label = UILabel()
            label.text = message
            label.textColor = .darkGray
            label.font = UIFont.systemFont(ofSize: 14)
            label.textAlignment = .center
            label.translatesAutoresizingMaskIntoConstraints = false
            
            container.addSubview(indicator)
            container.addSubview(label)
            overlay.addSubview(container)
            
            NSLayoutConstraint.activate([
                container.centerXAnchor.constraint(equalTo: overlay.centerXAnchor),
                container.centerYAnchor.constraint(equalTo: overlay.centerYAnchor),
                container.widthAnchor.constraint(greaterThanOrEqualToConstant: 120),
                container.heightAnchor.constraint(greaterThanOrEqualToConstant: 100),
                
                indicator.topAnchor.constraint(equalTo: container.topAnchor, constant: 20),
                indicator.centerXAnchor.constraint(equalTo: container.centerXAnchor),
                
                label.topAnchor.constraint(equalTo: indicator.bottomAnchor, constant: 12),
                label.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 16),
                label.trailingAnchor.constraint(equalTo: container.trailingAnchor, constant: -16),
                label.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: -16)
            ])
            
            vc.view.addSubview(overlay)
            self?.loadingView = overlay
            
            callback(.success(nil))
        }
    }
    
    private func hideLoading(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.loadingView?.removeFromSuperview()
            self?.loadingView = nil
            callback(.success(nil))
        }
    }
    
    // MARK: - ActionSheet
    
    private func showActionSheet(params: [String: AnyCodable], callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let title = params["title"]?.value as? String
        let options = params["options"]?.value as? [String] ?? []
        let cancelText = params["cancelText"]?.value as? String ?? "取消"
        
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.viewController else {
                callback(.failure(BridgeError.unknown("无法获取视图控制器")))
                return
            }
            
            let actionSheet = UIAlertController(title: title, message: nil, preferredStyle: .actionSheet)
            
            for (index, option) in options.enumerated() {
                actionSheet.addAction(UIAlertAction(title: option, style: .default) { _ in
                    callback(.success([
                        "index": index,
                        "option": option,
                        "cancelled": false
                    ]))
                })
            }
            
            actionSheet.addAction(UIAlertAction(title: cancelText, style: .cancel) { _ in
                callback(.success([
                    "index": -1,
                    "option": "",
                    "cancelled": true
                ]))
            })
            
            // iPad 适配
            if let popover = actionSheet.popoverPresentationController {
                popover.sourceView = vc.view
                popover.sourceRect = CGRect(x: vc.view.bounds.midX, y: vc.view.bounds.midY, width: 0, height: 0)
                popover.permittedArrowDirections = []
            }
            
            vc.present(actionSheet, animated: true)
        }
    }
}

File: example/android/app/src/main/java/com/aspect/webviewbridge/demo/modules/CustomModule.kt
/**
 * Custom 模块 - 自定义模块示例
 *
 * 演示如何创建自定义原生模块，供开发者参考
 * 包含 Alert、Confirm、Prompt、Toast 等常见 UI 交互方法
 */

package com.aspect.webviewbridge.demo.modules

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.EditText
import android.widget.Toast
import com.aspect.webviewbridge.protocol.*

/**
 * 自定义模块 - 演示原生 UI 交互
 * 
 * 此模块展示如何创建自定义模块，包含：
 * - Alert: 显示原生警告框
 * - Confirm: 显示确认对话框
 * - Prompt: 显示输入对话框  
 * - Toast: 显示 Toast 消息
 * - Loading: 显示/隐藏加载指示器
 */
class CustomModule(
    private val context: Context,
    private val bridgeContext: BridgeModuleContext,
    private val activityProvider: () -> Activity?
) : BridgeModule {

    override val moduleName = "Custom"

    override val methods = listOf(
        "Alert",
        "Confirm",
        "Prompt",
        "Toast",
        "ShowLoading",
        "HideLoading",
        "ActionSheet"
    )
    
    private val mainHandler = Handler(Looper.getMainLooper())
    
    // 加载对话框
    private var loadingDialog: AlertDialog? = null

    override fun handleRequest(
        method: String,
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        when (method) {
            "Alert" -> showAlert(request, callback)
            "Confirm" -> showConfirm(request, callback)
            "Prompt" -> showPrompt(request, callback)
            "Toast" -> showToast(request, callback)
            "ShowLoading" -> showLoading(request, callback)
            "HideLoading" -> hideLoading(callback)
            "ActionSheet" -> showActionSheet(request, callback)
            else -> callback(Result.failure(BridgeError.methodNotFound("$moduleName.$method")))
        }
    }

    // MARK: - Alert
    
    /**
     * 显示原生警告框
     * 
     * 参数：
     * - title: 标题
     * - message: 消息内容
     * - buttonText: 按钮文字（默认"确定"）
     */
    private fun showAlert(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val title = request.getString("title") ?: ""
        val message = request.getString("message") ?: ""
        val buttonText = request.getString("buttonText") ?: "确定"
        
        val activity = activityProvider()
        if (activity == null) {
            callback(Result.failure(BridgeError.internalError("无法获取 Activity")))
            return
        }
        
        mainHandler.post {
            try {
                AlertDialog.Builder(activity)
                    .setTitle(title)
                    .setMessage(message)
                    .setPositiveButton(buttonText) { dialog, _ ->
                        dialog.dismiss()
                        callback(Result.success(mapOf("action" to "confirm")))
                    }
                    .setCancelable(false)
                    .show()
            } catch (e: Exception) {
                callback(Result.failure(BridgeError.internalError(e.message)))
            }
        }
    }

    // MARK: - Confirm
    
    /**
     * 显示确认对话框
     * 
     * 参数：
     * - title: 标题
     * - message: 消息内容
     * - confirmText: 确认按钮文字（默认"确定"）
     * - cancelText: 取消按钮文字（默认"取消"）
     * 
     * 返回：
     * - confirmed: 是否点击了确认
     */
    private fun showConfirm(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val title = request.getString("title") ?: ""
        val message = request.getString("message") ?: ""
        val confirmText = request.getString("confirmText") ?: "确定"
        val cancelText = request.getString("cancelText") ?: "取消"
        
        val activity = activityProvider()
        if (activity == null) {
            callback(Result.failure(BridgeError.internalError("无法获取 Activity")))
            return
        }
        
        mainHandler.post {
            try {
                AlertDialog.Builder(activity)
                    .setTitle(title)
                    .setMessage(message)
                    .setPositiveButton(confirmText) { dialog, _ ->
                        dialog.dismiss()
                        callback(Result.success(mapOf(
                            "confirmed" to true,
                            "action" to "confirm"
                        )))
                    }
                    .setNegativeButton(cancelText) { dialog, _ ->
                        dialog.dismiss()
                        callback(Result.success(mapOf(
                            "confirmed" to false,
                            "action" to "cancel"
                        )))
                    }
                    .setCancelable(false)
                    .show()
            } catch (e: Exception) {
                callback(Result.failure(BridgeError.internalError(e.message)))
            }
        }
    }

    // MARK: - Prompt
    
    /**
     * 显示输入对话框
     * 
     * 参数：
     * - title: 标题
     * - message: 消息内容
     * - placeholder: 输入框占位符
     * - defaultValue: 默认值
     * - confirmText: 确认按钮文字
     * - cancelText: 取消按钮文字
     * 
     * 返回：
     * - confirmed: 是否点击了确认
     * - value: 输入的值（仅确认时有值）
     */
    private fun showPrompt(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val title = request.getString("title") ?: ""
        val message = request.getString("message") ?: ""
        val placeholder = request.getString("placeholder") ?: ""
        val defaultValue = request.getString("defaultValue") ?: ""
        val confirmText = request.getString("confirmText") ?: "确定"
        val cancelText = request.getString("cancelText") ?: "取消"
        
        val activity = activityProvider()
        if (activity == null) {
            callback(Result.failure(BridgeError.internalError("无法获取 Activity")))
            return
        }
        
        mainHandler.post {
            try {
                val input = EditText(activity).apply {
                    hint = placeholder
                    setText(defaultValue)
                    setPadding(48, 32, 48, 32)
                }
                
                AlertDialog.Builder(activity)
                    .setTitle(title)
                    .setMessage(message)
                    .setView(input)
                    .setPositiveButton(confirmText) { dialog, _ ->
                        dialog.dismiss()
                        callback(Result.success(mapOf(
                            "confirmed" to true,
                            "action" to "confirm",
                            "value" to input.text.toString()
                        )))
                    }
                    .setNegativeButton(cancelText) { dialog, _ ->
                        dialog.dismiss()
                        callback(Result.success(mapOf(
                            "confirmed" to false,
                            "action" to "cancel",
                            "value" to null
                        )))
                    }
                    .setCancelable(false)
                    .show()
            } catch (e: Exception) {
                callback(Result.failure(BridgeError.internalError(e.message)))
            }
        }
    }

    // MARK: - Toast
    
    /**
     * 显示 Toast 消息
     * 
     * 参数：
     * - message: 消息内容
     * - duration: 显示时长 "short" | "long"
     */
    private fun showToast(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val message = request.getString("message") ?: ""
        val duration = request.getString("duration") ?: "short"
        
        val toastDuration = if (duration == "long") Toast.LENGTH_LONG else Toast.LENGTH_SHORT
        
        mainHandler.post {
            Toast.makeText(context, message, toastDuration).show()
        }
        
        callback(Result.success(null))
    }

    // MARK: - ShowLoading
    
    /**
     * 显示加载指示器
     * 
     * 参数：
     * - message: 加载提示文字（默认"加载中..."）
     */
    private fun showLoading(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val message = request.getString("message") ?: "加载中..."
        
        val activity = activityProvider()
        if (activity == null) {
            callback(Result.failure(BridgeError.internalError("无法获取 Activity")))
            return
        }
        
        mainHandler.post {
            try {
                // 先关闭现有的
                loadingDialog?.dismiss()
                
                loadingDialog = AlertDialog.Builder(activity)
                    .setMessage(message)
                    .setCancelable(false)
                    .create()
                
                loadingDialog?.show()
                callback(Result.success(null))
            } catch (e: Exception) {
                callback(Result.failure(BridgeError.internalError(e.message)))
            }
        }
    }

    // MARK: - HideLoading
    
    /**
     * 隐藏加载指示器
     */
    private fun hideLoading(callback: (Result<Any?>) -> Unit) {
        mainHandler.post {
            loadingDialog?.dismiss()
            loadingDialog = null
        }
        callback(Result.success(null))
    }

    // MARK: - ActionSheet
    
    /**
     * 显示操作表
     * 
     * 参数：
     * - title: 标题（可选）
     * - options: 选项数组
     * - cancelText: 取消按钮文字（默认"取消"）
     * 
     * 返回：
     * - index: 选中的索引（取消时为 -1）
     * - option: 选中的选项文字
     */
    private fun showActionSheet(request: BridgeRequest, callback: (Result<Any?>) -> Unit) {
        val title = request.getString("title")
        val options = request.getStringArray("options") ?: emptyList()
        val cancelText = request.getString("cancelText") ?: "取消"
        
        if (options.isEmpty()) {
            callback(Result.failure(BridgeError.invalidParams("options 不能为空")))
            return
        }
        
        val activity = activityProvider()
        if (activity == null) {
            callback(Result.failure(BridgeError.internalError("无法获取 Activity")))
            return
        }
        
        mainHandler.post {
            try {
                val builder = AlertDialog.Builder(activity)
                
                if (!title.isNullOrEmpty()) {
                    builder.setTitle(title)
                }
                
                builder.setItems(options.toTypedArray()) { dialog, which ->
                    dialog.dismiss()
                    callback(Result.success(mapOf(
                        "index" to which,
                        "option" to options[which],
                        "cancelled" to false
                    )))
                }
                
                builder.setNegativeButton(cancelText) { dialog, _ ->
                    dialog.dismiss()
                    callback(Result.success(mapOf(
                        "index" to -1,
                        "option" to null,
                        "cancelled" to true
                    )))
                }
                
                builder.show()
            } catch (e: Exception) {
                callback(Result.failure(BridgeError.internalError(e.message)))
            }
        }
    }
}
